### DISCUSSION TITLE
Thoughts and questions on "EEG Box to PC Streaming Protocol" (on wiki)

### DISCUSSION THREAD

**Discussion by rtullybarr** [VolksEEG/VolksEEG#16] [Ideas]
Comments and questions pertain to this page sketching ideas for the PC to embedded device communication protocol:  https://github.com/VolksEEG/VolksEEG/wiki/EEG-Box-to-PC-Streaming-Protocol  

1. Do we require a synchronization triplet as described in the first section, if we use a framing protocol like the one Graham is proposing? Or is that an either/or question?
2. Yay version numbers!
3. How do we want to communicate what command is intended by the PC software? One idea (which I'm poaching from a similar API I've been exposed to) would use a byte or two to identify the command, and might have extra fields (fixed or variable length)

Possible examples:
0xA5 (start byte)
0x01 (version)
0x0002 (length of payload)
0x01 - command group: 'query settings'
0x01 - subcommand: 'query sample rate'
0x1234 (checksum - not a real one :P)

0xA5 (start byte)
0x01 (version)
0x0002 (length of payload)
0x02 - command group: 'write settings'
0x01 - subcommand: 'set sample rate'
0x01F4 - set sampling rate to 500 samples (per second is implied?)
0x1234 (checksum)

I like the idea of grouping related commands by using the same 'command group' byte, but this isn't necessary (and may or may not be useful). If we choose to do this, we can talk about what's the most logical way to group them (e.g. I've grouped 'querying settings' together here, but we could just as well give 'sample rate' one command group and have 'query sample rate' and 'set sample rate' be sub-commands within that).

4. Some commands issued by the PC will receive a response - querying settings, for example. It seems to me that it will make sense for most commands, possibly all, to have an 'acknowledgement' even if there's no data to be communicated back (and possibly implement retry logic if it's not received? Timeout / max retries?)

5. Do we want this to be fully driven by the PC software (e.g. for retrieving data,)? My sense is that we will want the data capture, at least, to be sent to the PC in 'unsolicited' messages, probably after receiving a command from the PC to enable capture.

6. Given that we'll have this structure, I imagine the channel data itself will be communicated in a command packet like these? Do we want to allow that to be a variable number of channels (as opposed to a fixed 8 channels?)

7. Maybe getting into scope creep, but I noticed we had a question about being compatible with Lab Streaming Layer (LSL): https://docs.openbci.com/Software/CompatibleThirdPartySoftware/LSL/. Would it be worthwhile to explore their protocols, and see if it might be appropriate to have both the EEG box and the PC software implement them and use those to communicate? In principle, that would seem to increase the flexibility of both - the embedded device could be used directly with other LSL-compatible software, and the PC software would be able to consume data from sources other than our device.

**Comment by alcohen** [comment]
These are great comments and questions @rtullybarr, basically an outline of the decisions we need to make and good suggestions regarding those decisions.

These decisions might be impacted by the ADS1299 functionality. It might make sense for a meeting with you, the two ADS1299 wizards (@longevity-software and Juan) and whoever else is interested? Would people have time later this week? Juan's not on GitHub yet so I have to ask him separately

**Reply by longevity-software** [reply]
@alcohen I think the functionality of the ADS1299 will be abstracted away at this point, I think we should implement what we want from the communications interface, and then translate it to/from ADS1299 language in the middle and abstraction layers.

**Comment by longevity-software** [comment]
Some good questions @rtullybarr, see my initial responses below.

> 1. Do we require a synchronization triplet as described in the first section, if we use a framing protocol like the one Graham is 
proposing? Or is that an either/or question?

I was proposing an alternative to @alcohens suggestion, the Start byte is essentially the same as the synchronisation triplet, so it would be an either or thing.

> 2. Yay version numbers!

Yay the yaying

> 3. How do we want to communicate what command is intended by the PC software? One idea (which I'm poaching from a similar API I've been exposed to) would use a byte or two to identify the command, and might have extra fields (fixed or variable length)

I like the idea of command groups, I have not seen that done before, but it seems like a neat way of organising things.

> 4. Some commands issued by the PC will receive a response - querying settings, for example. It seems to me that it will make sense for most commands, possibly all, to have an 'acknowledgement' even if there's no data to be communicated back (and possibly implement retry logic if it's not received? Timeout / max retries?)

Agree with that, although see question 5.

> 5. Do we want this to be fully driven by the PC software (e.g. for retrieving data,)? My sense is that we will want the data capture, at least, to be sent to the PC in 'unsolicited' messages, probably after receiving a command from the PC to enable capture.

Either way could be done, the PC could request the number of samples since it last requested data, or the firmware could push up new data points as and when it captures them. 

having two masters is always more complicated to implement, and if the firmware is master for pushing sample data and a packet is lost/corrupt between the firmware and PC, what do we do? does the firmware resend that sample as a retry or does the PC request specific data points? either way the PC has to deal with out of sequence data points, or wait for the data point to be retried/re-requested.

If each data point has an index, the PC could request newer data points than the last index that it has received, missing data points would then just be dealt with naturally by retrying this process.  

> 6. Given that we'll have this structure, I imagine the channel data itself will be communicated in a command packet like these? Do we want to allow that to be a variable number of channels (as opposed to a fixed 8 channels?)

A variable number of channels would make it more flexible for moving to higher channel numbers, and also allow for lower channel counts at the same time.

On the system I've worked on previously, we could turn channels on and off, so, even for an 8 channel system there was always the possibility of a lower channel count and we implemented a variable channel packet for that.

Given that each sample will be the same size and we can't turn off channels (@alcohen would turning channels off be a requirement?), we could use something similar to this within the data packet. 

1) number of channels being transmitted
2) first channel value
3) second channel value
4) ....... channel value etc etc

Alternatively it may be beneficial to add a "number of samples" value to allow more than 1 sample per packet and a channel identifier before each channels samples, similar to this:

1) number of channels being transmitted
2) number of samples per channel
3) channel identifier 0
4 to (n+4)) channel 0's samples
n + 5) channel identifier 1
6 to (n + 6)) channel 1's samples
etc etc.

> 7. Maybe getting into scope creep, but I noticed we had a question about being compatible with Lab Streaming Layer (LSL): https://docs.openbci.com/Software/CompatibleThirdPartySoftware/LSL/. Would it be worthwhile to explore their protocols, and see if it might be appropriate to have both the EEG box and the PC software implement them and use those to communicate? In principle, that would seem to increase the flexibility of both - the embedded device could be used directly with other LSL-compatible software, and the PC software would be able to consume data from sources other than our device.

I also noticed the question about LSL, but have not looked into it yet, if it provides us with everything we need then it makes sense to use something which is already out there.

**Reply by rtullybarr** [reply]
> Alternatively it may be beneficial to add a "number of samples" value to allow more than 1 sample per packet and a channel identifier   
> before each channels samples, similar to this:

>    number of channels being transmitted
>    number of samples per channel
>    channel identifier 0
>    4 to (n+4)) channel 0's samples
>    n + 5) channel identifier 1
>    6 to (n + 6)) channel 1's samples
>    etc etc.

I hadn't thought of it, but I like having a channel identifier as in this example, even if we don't transmit multiple samples. It would let us support turning channels 'on/off' and removes any possible ambiguity with determining the 'channel number' by position within the packet.
Allowing multiple samples per packet does seem useful to me - if we do have acknowledgements for all messages, this would prevent the round-trip time of message+response from letting us get 'behind' on retrieving the available data.

@alcohen and all, I should be available for a meeting next weekend mornings, or Friday after 11am PST
